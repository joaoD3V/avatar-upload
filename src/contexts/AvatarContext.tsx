import { createContext, useState, useCallback } from 'react';

export type AvatarStatus = 'initial' | 'editing' | 'saved' | 'error';

export type ImageFileProps = string | ArrayBuffer | undefined;

type AvatarContexType = {
  status: AvatarStatus;
  avatarImage: ImageFileProps;
  sliderValue: number;
  handleClose: () => void;
  handleChangeStatus: (newStatus: AvatarStatus) => void;
  handlePreviewProfileImage: (files: File[]) => void;
  handleUpdateSliderValue: (value: number | number[]) => void;
};

type AvatarContextProviderProps = {
  children: React.ReactNode;
};

export const AvatarContext = createContext({} as AvatarContexType);

export function AvatarContextProvider({
  children,
}: AvatarContextProviderProps) {
  const [status, setStatus] = useState<AvatarStatus>('initial');
  const [avatarImage, setAvatarImage] = useState<ImageFileProps>(undefined);
  const [sliderValue, setSliderValue] = useState<number>(0);

  const handleClose = useCallback(() => {
    setStatus('initial');
    setAvatarImage(undefined);
  }, []);

  const handleChangeStatus = useCallback((newStatus: AvatarStatus) => {
    setStatus(newStatus);
  }, []);

  const handlePreviewProfileImage = useCallback((files: File[]) => {
    if (files && files[0]) {
      const reader = new FileReader();
      reader.addEventListener('load', (event) => {
        const image = event.target!.result!;

        const imageCopy = new Image();
        imageCopy.src = `${image}`;

        imageCopy.onload = () => {
          setAvatarImage(image);
          setSliderValue(0);
          return setStatus('editing');
        };
      });

      reader.readAsDataURL(files[0]);
    }
  }, []);

  const handleUpdateSliderValue = useCallback((value: number | number[]) => {
    setSliderValue(Number(value));
  }, []);

  return (
    <AvatarContext.Provider
      value={{
        status,
        avatarImage,
        sliderValue,
        handleClose,
        handleChangeStatus,
        handlePreviewProfileImage,
        handleUpdateSliderValue,
      }}
    >
      {children}
    </AvatarContext.Provider>
  );
}
