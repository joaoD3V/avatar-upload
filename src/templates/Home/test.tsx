import 'jest-canvas-mock';
import { screen } from '@testing-library/react';
import { AvatarContext, AvatarContextProvider } from 'contexts/AvatarContext';
import { renderWithTheme } from 'utils/tests/helpers';

import { Home } from '.';

describe('<Home />', () => {
  it('should render top-level wrapper correctly', () => {
    const { container } = renderWithTheme(<Home />);

    expect(screen.getByLabelText(/Home template/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render AvatarUpload when status is initial', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('initial')}
              <Home />
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.queryByLabelText(/close button/i)).not.toBeInTheDocument();
    expect(screen.getByLabelText(/wrapper/i)).toBeInTheDocument();
    expect(screen.queryByLabelText(/Avatar canva/i)).not.toBeInTheDocument();
    expect(
      screen.getByAltText(
        /Illustration of mountains and sun, symbolizing a picture icon/i
      )
    ).toBeInTheDocument();
    expect(screen.getByText(/Organization Logo/i)).toBeInTheDocument();
    expect(
      screen.getByText(/Drop the image here or click to browse./i)
    ).toBeInTheDocument();
    expect(screen.getByTestId(/drop-input/i)).toBeInTheDocument();
    expect(screen.getByTestId(/drop-wrapper/i)).toBeInTheDocument();
  });

  it('should not render AvatarEditing when status is editing and avatarImage is undefined', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('editing')}
              <Home />
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.queryByLabelText(/Avatar canva/i)).not.toBeInTheDocument();
    expect(screen.queryByLabelText(/Crop slider/i)).not.toBeInTheDocument();
    expect(
      screen.queryByRole('button', { name: 'Save' })
    ).not.toBeInTheDocument();
  });

  it('should render AvatarError when status is error', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('error')}
              <Home />
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(
      screen.getByRole('img', { name: 'Exclamation icon' })
    ).toBeInTheDocument();
    expect(screen.getByText(/Sorry, the upload failed./i)).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: 'Try again' })
    ).toBeInTheDocument();
  });
});
