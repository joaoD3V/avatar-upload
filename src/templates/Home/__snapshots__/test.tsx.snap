// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render top-level wrapper correctly 1`] = `
.c0 {
  width: 100vw;
  height: 100vh;
  background: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<main
  aria-label="Home template"
  class="c0"
/>
`;
